ganttproject/src/main/java/net/sourceforge/ganttproject/chart/UnboundedViewportTaskSceneBuilder.kt

```diff
@@ -0,0 +1,184 @@
+    val allTasks = tasks.flatMap { it.flatten() }
+    paintTasks(allTasks, canvas)
+    paintDependencies(allTasks, canvas)
+    return canvas
+  }
+
+  private fun paintTasks(tasks: List<Task>, canvas: Canvas) {
+    val taskBarOffset = (input.rowHeight - input.barHeight) / 2
+    val taskActivityScene = TaskActivitySceneBuilder(
+      ActivitySceneTaskApi, input, canvas, TaskLabelSceneBuilderStub, TaskActivitySceneBuilder.Style(taskBarOffset)
+    )
+    tasks.forEachIndexed { row, task ->
+      taskActivityScene.renderActivities(row, listOf(task.activity), input.bottomUnitOffsets)
+      val nextLineY = row * input.rowHeight
+      val nextLine = canvas.createLine(0, nextLineY, input.viewportWidth, nextLineY)
+      nextLine.foregroundColor = Color.GRAY
+    }
+  }
+
+  private fun paintDependencies(tasks: List<Task>, canvas: Canvas) {
+    val dependencyRenderer = DependencySceneBuilder(
+      canvas, canvas.newLayer(), DependencySceneTaskApi(tasks, logger), { input.barHeight }
+    )
+    dependencyRenderer.build()
+  }
+
+  abstract class Task {
+    abstract val start: Date
+    abstract val end: Date
+    abstract val isMilestone: Boolean
+    abstract val subtasks: List<Task>
+    abstract val dependants: List<Dependant>
+
+    val activity = Activity()
+
+    fun flatten(result: MutableList<Task> = mutableListOf()): List<Task> {
+      result.add(this)
+      subtasks.forEach { it.flatten(result) }
+      return result
+    }
+
+    inner class Activity : BarChartActivity<Task> {
+      override fun getStart() = this@Task.start
+      override fun getEnd() = this@Task.end
+      override fun getOwner() = this@Task
+
+      override fun getDuration(): TimeDuration {
+        return TimeDurationImpl(
+          GPTimeUnitStack.DAY,
+          TimeUnit.DAYS.convert(start.time - end.time, TimeUnit.MILLISECONDS)
+        )
+      }
+    }
+
+    class Dependant(val task: Task, val type: TaskDependencyConstraint.Type, val hardness: Hardness)
+  }
+
+  private class DependencySceneTaskApi(
+    private val tasks: List<Task>,
+    private val logger: ((String) -> Unit)?
+  ) : DependencySceneBuilder.TaskApi<Task, TaskDependency> {
+    override fun isMilestone(task: Task) = task.isMilestone
+    override fun getTasks() = tasks
+
+    override fun getUnitVector(activity: BarChartActivity<Task>, connector: TaskDependency): Dimension? {
+      return when (activity) {
+        connector.start -> connector.startVector
+        connector.end -> connector.endVector
+        else -> null.also { logger?.invoke("Connector-activity mismatch") }
+      }
+    }
+
+    override fun getStyle(dependency: TaskDependency): String {
+      return if (dependency.dependant.hardness === Hardness.STRONG) "dependency.line.hard" else "dependency.line.rubber"
+    }
+
+    override fun getConnectors(task: Task): Iterable<TaskDependency> {
+      return task.dependants.map { TaskDependency(task, it) }
```

Comment:
Is it a typo or as designed? The prototype code in Java calls `task.getDependencies()` (all deps, not only dependants).

Reply:
It's intentional, this way we go through all dependencies anyway. I tested it on HouseBuildingSample.gan and it works as expected. Looks like the original code does extra work here or I just haven't notice some special cases.